src/repositories/job-posting-repository.ts

```diff
@@ -0,0 +1,83 @@
+import { JobPostings, JobPostingsData } from '@/models/job-postings-model';
+import { Transaction } from 'objection';
+
+/**
+ * Interface for JobPostingRepository
+ */
+export interface IJobPostingRepository {
+  findJobPostingByUuid(jobPostingUuid: string): Promise<JobPostingsData | null>;
+  findWithActiveApplication(jobPostingUuid: string): Promise<boolean>;
+  softDelete(jobPostingUuid: string, trx?: Transaction): Promise<number>;
+  restore(jobPostingUuid: string, trx?: Transaction): Promise<number>;
+}
+
+/**
+ * Repository for job_postings table operations, focused on deletion and restoration logic.
+ */
+export class JobPostingRepository implements IJobPostingRepository {
+  /**
+   * Private method to find job posting by UUID with optional field selection
+   * @param jobPostingUuid job posting uuid
+   * @param selectFields optional array of fields to select
+   * @returns JobPostingsData or null
+   */
+  private async findJobPostingByUuidWithSelect(jobPostingUuid: string, selectFields?: string[]) {
+    const query = JobPostings.query().findOne({ uuid: jobPostingUuid });
+    if (selectFields && selectFields.length > 0) {
+      query.select(...selectFields);
+    }
+    return await query;
+  }
+
+  /**
+   * Find a job posting by its uuid
+   * @param jobPostingUuid job posting uuid
+   * @returns JobPostingsData or null
+   */
+  async findJobPostingByUuid(jobPostingUuid: string): Promise<JobPostingsData | null> {
+    const job = await this.findJobPostingByUuidWithSelect(jobPostingUuid);
+    return job ?? null;
+  }
+
+  /**
+   * Efficiently check if a job posting has any application with status 'pending' or 'under_review' using .first()
+   * @param jobPostingUuid job posting uuid
+   * @returns true if there is at least one such application, false otherwise
+   */
+  async findWithActiveApplication(jobPostingUuid: string): Promise<boolean> {
+    const job = await this.findJobPostingByUuidWithSelect(jobPostingUuid, ['id', 'uuid']);
+    if (!job) return false;
+    
+    const application = await JobPostings.relatedQuery('jobApplications')
+      .for(job.id)
+      .whereIn('status', ['pending', 'under_review'])
+      .first();
+    return !!application;
+  }
```

Comment:
The current implementation makes two separate database queries: one to get the job post and another to find related applications. This can be done in a single, more efficient query using whereExists.

**Reasoning**: You don't actually need any data from the JobPostings table itself, you just need to know if a record exists that matches both the UUID and the application status criteria. A subquery is perfect for this.